net/instaweb/rewriter/csp.cc

```diff
@@ -172,12 +175,158 @@ CspSourceExpression CspSourceExpression::Parse(StringPiece input) {
+  // TODO(morlovich): This means redirect following may require changes
+  // here.
+  if (kind_ != kSelf && kind_ != kSchemeSource && kind_ != kHostSource) {
+    return false;
+  }
+
+  if (!origin_url.IsAnyValid() || !url.IsAnyValid()) {
+    return false;
+  }
+
+  // Check for 'self' first, since that doesn't need/have url_data()
+  if (kind_ == kSelf) {
+    if (origin_url.Origin() == url.Origin()) {
+      return true;
+    }
+
+    if (origin_url.Host() != url.Host()) {
+      return false;
+    }
+
+    if (origin_url.SchemeIs("http") && url.SchemeIs("https")) {
+      // Using the same port is OK.
+      if (origin_url.EffectiveIntPort() == url.EffectiveIntPort()) {
+        return true;
+      }
+
+      // Using default ports for both is OK, too.
+      if (HasDefaultPortForScheme(origin_url) && HasDefaultPortForScheme(url)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  // Give our state some short names closer to those in the spec
+  StringPiece expr_scheme = url_data().scheme_part;
+  StringPiece expr_host = url_data().host_part;
+  StringPiece expr_port = url_data().port_part;
+  const std::vector<GoogleString>& expr_path = url_data().path_part;
+
+  // Some special handling of *, which for some reason handles some schemes
+  // a bit differently than other things with * host portion and no scheme
+  // specified.
+  if (kind_ == kHostSource &&
+      expr_scheme.empty() &&
+      expr_host == "*" &&
+      expr_port.empty() &&
+      expr_path.empty()) {
+    if (url.SchemeIs("http") ||
+        url.SchemeIs("https") ||
+        url.SchemeIs("ftp")) {
+      return true;
+    }
+    return (url.Scheme() == origin_url.Scheme());
+  }
+
+  if (!expr_scheme.empty()
+      && url.Scheme() != expr_scheme
+      && !(expr_scheme == "http" && url.SchemeIs("https"))) {
+    return false;
+  }
+
+  if (kind_ == kSchemeSource) {
+    return true;
+  }
+
+  if (url.Host().empty() || expr_host.empty()) {
+    return false;
+  }
+
+  if (expr_scheme.empty()
+      && url.Scheme() != origin_url.Scheme()
+      && !(origin_url.SchemeIs("http") && url.SchemeIs("https"))) {
+    return false;
+  }
+
+  if (expr_host[0] == '*') {
+    StringPiece remaining = expr_host.substr(1);
+    if (!url.Host().ends_with(remaining)) {
+      return false;
+    }
+  } else {
+    if (url.Host() != expr_host) {
+      return false;
+    }
+  }
+
+  // TODO(morlovich): Implement IP-address handling here, once appropriate
+  // spec has been read.
+
+  if (expr_port.empty()) {
+    if (!HasDefaultPortForScheme(url)) {
+      return false;
+    }
+  } else {
+    // TODO(morlovich): Check whether the :80/:443 case is about effective
+    // or explicit port.
+    if (expr_port != "*"
+        && expr_port != IntegerToString(url.EffectiveIntPort())
+        && !(expr_port == "80" && url.EffectiveIntPort() == 443)) {
+      return false;
+    }
+  }
+
+  if (!expr_path.empty()) {  // this would also be skipped for redirects
```

Comment:
re: "This is the only spot that will need to be changed for redirects --- basically steps after redirect don't do the path check." 
Ok, thanks for clearing that up